### PR TITLE
Allow testing without going through a relay

### PR DESCRIPTION
**Pull request by jessetane**
possible fix for #124

cc @KaptenJansson

### CODE REVIEW

**Comment by jessetane** [comment]
@googlebot I signed it!

**Comment by KaptenJansson** [comment]
Thank you for the pull request! Will take a look and comment inline.

File: bower.json [hunk 1]
```diff
@@ -15,7 +15,7 @@
   ],
   "license": "BSD-3-Clause",
   "dependencies": {
-    "polymer": "Polymer/polymer#1.0.9",
+    "polymer": "Polymer/polymer#1.1.2",
```
Comment (KaptenJansson) [inline bower.json]: Is there a specific reason for bumping the version?

File: src/js/call.js [hunk 2]
```diff
@@ -127,6 +127,14 @@ Call.isIpv6 = function(candidate) {
   return candidate.address.indexOf(':') !== -1;
 };
 
+Call.isUDP = function(candidate) {
```
Comment (KaptenJansson) [inline src/js/call.js]: Could you change this to lower case to match the others? I.e Call.isUdp.

File: src/js/call.js [hunk 3]
```diff
@@ -127,6 +127,14 @@ Call.isIpv6 = function(candidate) {
...
+  return candidate.protocol === 'udp';
+};
+
+Call.isTCP = function(candidate) {
```
Comment (KaptenJansson) [inline src/js/call.js]: Same as above.

**Comment by KaptenJansson** [comment]
Could you also update the text in the settings to reflect that STUN servers can be used now?

LGTM apart from the nits mentioned.

**Comment by KaptenJansson** [comment]
Just realized one potential issue here, by using STUN we cannot effectively check if we have reached the outside world as the TCP and UDP connectivity tests grabs the first candidate, and if it's a HOST candidate, it's not a useful indicator for if you have reached the STUN server.

**Comment by KaptenJansson** [comment]
What I forgot to mention about my last comment is that we can ofc add more checks to tests to see if we get a reflexive candidate and then pass, just keep in mind that we will not always get a reflexive candidates if you can reach the STUN server without traversing any NAT's.

File: bower.json (hunk 1 above)
Comment (jessetane) [inline bower.json]: the automated tests were failing because other deps aren't pinned:

https://travis-ci.org/webrtc/testrtc/jobs/82963114

Comment (KaptenJansson) [inline bower.json]: Aha, thanks for the details.

**Comment by KaptenJansson** [comment]
Yes it is, thanks!

Have you read my last two comments about STUN and the tests? It's not that bad in reality as the default config is TURN and people entering a STUN server instead hopefully knows what they are doing.

**Comment by jessetane** [comment]
Yes, just thinking about it - I wanted to be able to test webrtc connectivity on a local network only without a relay (which seems as important as being able to reach the outside world, no?). Here is the original context:

https://github.com/feross/simple-peer/issues/41

**Comment by KaptenJansson** [comment]
Yes it is important, the problem is that HOST candidates are generated locally and not by the STUN server. If they are of type RELAY, you know that you have reached the TURN server. See the dilemma?

**Comment by jessetane** [comment]
I do, in my case it wasn't so much that I wanted to test connectivity as prove the lack of it! Host candidates were being generated properly and connectivity was still not being established.

**Comment by KaptenJansson** [comment]
Also one good page to check out what candidates are collected, try https://webrtc.github.io/samples/src/content/peerconnection/trickle-ice/.

**Comment by jessetane** [comment]
Perhaps what I wanted then, was not to use STUN but just to clear the default TURN servers...

**Comment by KaptenJansson** [comment]
By entering a new TURN server URL, it will use it. But this means the TURN server has to be reachable and if you want to test on a restricted network that does not have access to the Google Turn server, you have to setup your own.

Keep in mind though that a STUN server can return REFLEXIVE candidates which should be IP address of the closest firewall/router. Details here https://www.webrtc-experiment.com/docs/STUN-or-TURN.html.
This will be an indication that you reached the STUN server.

Also another thing, most TURN servers has STUN functionality as well, e.g return REFLEXIVE candidates.

**Comment by KaptenJansson** [comment]
To conclude, if you want testrtc to check STUN servers, we need to add tests that checks for REFLEXIVE candidates.

Is this something you want to do in a separate PR? Else I can take a stab at it.

**Comment by jessetane** [comment]
Yes, thanks for the clarifications - and as you mention we could easily force reflexive candidates if there is a STUN URI passed in, and maybe that makes sense, but it still leaves no way to test host candidates, which is I think what I was really trying to do.

**Comment by KaptenJansson** [comment]
But host candidates are generated by the client, i.e. checks if the network interface works basically, is that what you want to test? Or do you want to check for a particular local IP address?

**Comment by jessetane** [comment]
Yes, the former. In the case of the issue at Sauce labs, something truly strange must have been going on as even host candidates, though being generated, were not providing connectivity. I wanted to be able to prove that with this test.

What do you think about forcing reflexive for testing STUN, but also providing a checkbox to disable forcing the default relay?

**Comment by jessetane** [comment]
Or maybe the checkbox is to allow host candidates?

**Comment by jessetane** [comment]
Another idea: what about radio buttons to select between 3 "modes"?

1) Relay (tests connectivity to the outside world, requires a TURN URI, the default)
2) Reflexive (tests the router on your link, requires a STUN URI)
3) Host (tests your nic, no extra infos required)

Let me know if I'm making any incorrect assumptions about what these things mean...

File: src/js/call.js [hunk 4]
```diff
@@ -127,6 +127,14 @@ Call.isIpv6 = function(candidate) {
   return candidate.address.indexOf(':') !== -1;
 };
 
+Call.isUdp = function(candidate) {
+  return candidate.protocol === 'udp';
```
Comment (KaptenJansson) [inline src/js/call.js]: Actually, we need to check for reflexive candidates here in order to verify that you have reached a STUN server, it's type `srflx` which actually stand for server reflexive candidates (provided by the stun server).

File: src/js/conntest.js [hunk 5]
```diff
@@ -20,7 +20,8 @@ function udpConnectivityTest() {
   Call.asyncCreateTurnConfig(
       function(config) {
         filterConfig(config, 'udp');
-        gatherCandidates(config, null, Call.isRelay);
+        var filter = Call.configHasStunURI(config) ? Call.isUdp : Call.isRelay;
```
Comment (KaptenJansson) [inline src/js/conntest.js]: Change this to Call.isSrflx instead.

File: src/js/conntest.js [hunk 6]
```diff
@@ -31,7 +32,8 @@ function tcpConnectivityTest() {
   Call.asyncCreateTurnConfig(
       function(config) {
         filterConfig(config, 'tcp');
-        gatherCandidates(config, null, Call.isRelay);
+        var filter = Call.configHasStunURI(config) ? Call.isTcp : Call.isRelay;
```
Comment (KaptenJansson) [inline src/js/conntest.js]: Remove Call.isTcp

File: src/js/call.js [hunk 7]
```diff
@@ -127,6 +127,14 @@ Call.isIpv6 = function(candidate) {
...
+};
+
+Call.isTcp = function(candidate) {
```
Comment (KaptenJansson) [inline src/js/call.js]: Remove this.

**Comment by KaptenJansson** [comment]
@andresusanopinto PTAL.

**Comment by andresusanopinto** [comment]
I think there is a few misconceptions here. 

"UDP/TCP connectivity" tests are about verifying a given endpoint has udp/tcp connectivity to the public internet and not whether the endpoint can gather udp/tcp candidates (actually afaik webrtc in chrome does not supports tcp connectivity between endpoints).

Those tests accomplish that by trying to get a response from a server out in the internet using either udp or tcp. More specifically in the current state, the tests verify if they can allocate a relay address by talking with a turn server with a specific protocol. E.g. "turn:lalala?protocol=tcp".
It sounds perfectly reasonable to me to switch that to use a stun server, but then the test should verify it has a server reflexive candidate, which can only be obtained by talking with the stun/turn server.

Regarding the "Data throughput" test it gets a bit more confusing... the test wants to gather information on how good a user connection is and it does that by achieve a loopback call that goes through the turn server (to force the data to go out of the local network and be back). What information do you want to gather from the test by using only a stun server and by not filtering the local candidates?

**Comment by jessetane** [comment]
I agree that trying to talk about protocol and throughput at the same time as connectivity confuses things, so maybe before dealing with that we can agree on what types of connectivity are possible and whether testrtc can / should be testing them? Assumptions below:

1) Relay connectivity
If candidates are limited to type "relay" and connectivity can be established, it means both candidates have access to the public internet and the designated TURN server is functioning properly. This is what is currently tested and is nice to have.

2) Reflexive connectivity
If candidates are limited to type "srflx" and connectivity can be established, it means both candidates are behind NATs and hole punching is working as advertised. This requires being behind a NAT and having access to a STUN server. This is not currently tested, but I think would definitely be nice to have as it's probably where many common connectivity problems arise.

3) Host connectivity
If candidates are limited to type "host" and connectivity can be established, it means both peers either have the same ip, are on the same private network, or both peers have publicly routable ips. In testrtc's case both peers would always have the same ip and this would merely confirm the user's network interface is functioning properly. This is not currently tested, but I think this scenario is still worth supporting to confirm for example that two apps talking via webrtc on the same machine don't need to hit a router or a relay.

Again, just stating the way I currently understand things to work... please correct me where I'm mistaken.

**Comment by jessetane** [comment]
closing in favor of #127

**Pull request closed without merging**